scripts/scmigrate/migrate_test.go

```diff
@@ -0,0 +1,160 @@
+	t.Helper()
+
+	batch := db.NewBatch()
+
+	for idx, val := range data {
+		payload, err := proto.Marshal(val.commit.ToProto())
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if err := batch.Set(makeKeyFromPrefix(prefixSeenCommit, int64(idx)), payload); err != nil {
+			t.Fatal(err)
+		}
+	}
+	if err := batch.WriteSync(); err != nil {
+		t.Fatal(err)
+	}
+	if err := batch.Close(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestMigrations(t *testing.T) {
+	t.Run("Sort", func(t *testing.T) {
+		t.Run("HandCraftedData", func(t *testing.T) {
+			testData := []toMigrate{
+				{commit: &types.Commit{Height: 100}},
+				{commit: &types.Commit{Height: 0}},
+				{commit: &types.Commit{Height: 8}},
+				{commit: &types.Commit{Height: 1}},
+			}
+
+			sortMigrations(testData)
+			assertWellOrderedMigrations(t, testData)
+		})
+		t.Run("RandomGeneratedData", func(t *testing.T) {
+			testData := []toMigrate{{commit: &types.Commit{Height: 0}}}
+
+			testData = appendRandomMigrations(testData, 10000)
+
+			sortMigrations(testData)
+			assertWellOrderedMigrations(t, testData)
+		})
+	})
+	t.Run("GetMigrationsToDelete", func(t *testing.T) {
+		for i := 1; i < 100; i++ {
+			data := appendRandomMigrations([]toMigrate{}, i)
+			toMigrate := getMigrationsToDelete(data)
+			if len(data) != len(toMigrate)+1 {
+				t.Fatalf("migration prep did not save one document [original=%d migrations=%d]", len(data), len(toMigrate))
+			}
+		}
+	})
+	t.Run("InvalidMigrations", func(t *testing.T) {
+		if _, err := makeToMigratge(nil); err == nil {
+			t.Fatal("should error for nil migrations")
+		}
+		if _, err := makeToMigratge([]byte{}); err == nil {
+			t.Fatal("should error for empty migrations")
+		}
+		if _, err := makeToMigratge([]byte("invalid")); err == nil {
+			t.Fatal("should error for empty migrations")
+		}
+	})
+
+	t.Run("GetSeenCommits", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		db := dbm.NewMemDB()
+		data := appendRandomMigrations([]toMigrate{}, 100)
+		insertTestData(t, db, data)
+		commits, err := getAllSeenCommits(ctx, db)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if len(commits) != len(data) {
+			t.Log("inputs", len(data))
+			t.Log("commits", len(commits))
+			t.Fatal("migrations not found in database")
+		}
+	})
+	t.Run("Integration", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		db := dbm.NewMemDB()
+		data := appendRandomMigrations([]toMigrate{}, 1000)
+		insertTestData(t, db, data)
+
+		for _, test := range []string{"Migration", "Idempotency"} {
+			// run the test twice to make sure that it's
+			// safe to rerun
+			t.Run(test, func(t *testing.T) {
+				if err := Migrate(ctx, db); err != nil {
+					t.Fatal(err)
+				}
+
+				post, err := getAllSeenCommits(ctx, db)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if len(post) != 1 {
+					t.Fatal("migration was not successful")
+				}
```

Comment:
Do we also want to assert that the remaining seen commit is the latest? i.e. has height 1000

Reply:
we assert elsewhere that the the sort is correct, so I think the test is correct without this assertion, though I'll play with it (there are also 1000 records, the values are random.)